File: Example/Example/Classes/Module/Home/HomeViewController.m
//
//  ViewController.m
//  Example
//
//  Created by wuyong on 17/2/16.
//  Copyright © 2018年 wuyong.site. All rights reserved.
//

#import "HomeViewController.h"

#pragma mark - HomeViewController

@interface HomeViewController () <FWViewController>

@property (nonatomic, strong) UIButton *loginButton;

@end

@implementation HomeViewController

#pragma mark - Lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // FIXME: hotfix
    self.edgesForExtendedLayout = UIRectEdgeNone;
    
    // TODO: feature
    UIButton *loginButton = [UIButton buttonWithType:UIButtonTypeSystem];
    self.loginButton = loginButton;
    [loginButton addTarget:self action:@selector(onLogin) forControlEvents:UIControlEventTouchUpInside];
    loginButton.frame = CGRectMake(0, 20, FWScreenWidth, 30);
    [self.view addSubview:loginButton];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self renderData];
}

- (void)renderData {
    #if APP_PROD
    NSString *envTitle = FWLocalizedString(@"envProd");
    #elif APP_TEST
    NSString *envTitle = FWLocalizedString(@"envTest");
    #else
    NSString *envTitle = FWLocalizedString(@"envDev");
    #endif
    self.fwBarTitle = [NSString stringWithFormat:@"%@ - %@", FWLocalizedString(@"homeTitle"), envTitle];
    
    if ([Mediator.userModule isLogin]) {
        [self.loginButton setTitle:FWLocalizedString(@"backTitle") forState:UIControlStateNormal];
    } else {
        [self.loginButton setTitle:FWLocalizedString(@"welcomeTitle") forState:UIControlStateNormal];
    }
}

#pragma mark - Action

- (void)onLogin {
    if ([Mediator.userModule isLogin]) return;
    
    FWWeakifySelf();
    [Mediator.userModule login:^{
        FWStrongifySelf();
        [self renderData];
    }];
}

@end

File: Example/Example/Classes/Module/Home/HomeViewController.swift
//
//  HomeViewController.swift
//  Example
//
//  Created by wuyong on 2021/3/19.
//  Copyright © 2021 site.wuyong. All rights reserved.
//

import Foundation

class HomeViewController: UIViewController, FWViewController {
    private lazy var loginButton: UIButton = {
        let button = UIButton(type: .system)
        button.frame = CGRect(x: 0, y: 20, width: FWScreenWidth, height: 30)
        button.addTarget(self, action: #selector(onLogin), for: .touchUpInside)
        return button
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // FIXME: hotfix
        edgesForExtendedLayout = .bottom
        // TODO: feature
        fwView.addSubview(loginButton)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        renderData()
    }
    
    func renderData() {
        #if APP_PRODUCTION
        let envTitle = FWLocalizedString("envProduction")
        #elseif APP_STAGING
        let envTitle = FWLocalizedString("envStaging")
        #elseif APP_TESTING
        let envTitle = FWLocalizedString("envTesting")
        #else
        let envTitle = FWLocalizedString("envDevelopment")
        #endif
        fwBarTitle = "\(FWLocalizedString("homeTitle")) - \(envTitle)"
        
        if Mediator.userModule.isLogin() {
            loginButton.setTitle(FWLocalizedString("backTitle"), for: .normal)
        } else {
            loginButton.setTitle(FWLocalizedString("welcomeTitle"), for: .normal)
        }
    }
    
    // MARK: - Action
    @objc func onLogin() {
        if Mediator.userModule.isLogin() { return }
        
        Mediator.userModule.login { [weak self] in
            self?.renderData()
        }
    }
}
